zkml/src/layers/add.rs

```diff
@@ -97,54 +197,490 @@ impl<N> OpInfo for Add<N> {
+        let mut output_claims = vec![Claim::new(last_claim.point.clone(), left_eval)];
+        let right_eval = match self.operand {
+            Some((ref op, _)) => {
+                let right_eval = op.evals_flat::<E>().into_mle().evaluate(&last_claim.point);
+                let mut claims = HashMap::new();
+                claims.insert(
+                    OPERAND_POLY_ID.to_string(),
+                    Claim::new(last_claim.point.clone(), right_eval),
+                );
+                // this claim gets verified by the PCS openings since it's a static one
+                prover.add_common_claims(node_id, claims)?;
+                right_eval
+            }
+            None => {
+                let right_eval = step_data.inputs[1]
+                    .get_data()
+                    .to_vec()
+                    .into_mle()
+                    .evaluate(&last_claim.point);
+                // this claims gets passed to the previous layer alongside the left one.
+                output_claims.push(Claim::new(last_claim.point.clone(), right_eval));
+                right_eval
+            }
+        };
+        prover.push_proof(
+            node_id,
+            LayerProof::Add(AddProof {
+                left_eval,
+                right_eval,
+            }),
+        );
+        Ok(output_claims)
+    }
+}
+
+impl<E, PCS> VerifiableCtx<E, PCS> for AddCtx
+where
+    E: ExtensionField,
+    E::BaseField: Serialize + DeserializeOwned,
+    E: Serialize + DeserializeOwned,
+    PCS: PolynomialCommitmentScheme<E>,
+{
+    type Proof = AddProof<E>;
+
+    fn verify<T: Transcript<E>>(
+        &self,
+        proof: &Self::Proof,
+        last_claims: &[&Claim<E>],
+        verifier: &mut Verifier<E, T, PCS>,
+        _shape_step: &ShapeStep,
+    ) -> anyhow::Result<Vec<Claim<E>>> {
+        ensure!(last_claims.len() == 1, "Add layer expects 1 claim");
+        let last_claim = last_claims[0];
+        // just making sure downsizing due to API of E is ok
+        ensure!((self.quant_info.left_scale() as u64) as Element == self.quant_info.left_scale());
+        ensure!((self.quant_info.right_scale() as u64) as Element == self.quant_info.right_scale());
+        // we have the output claim f(r) = y = x1(r) * x1_scale + x2(r) * x2_scale
+        // and the proof gives us x1(r) and x2(r) so we just need to "scale" these and
+        // verify the equation.
+        let left_scale: E = self.quant_info.left_scale().to_field();
+        let scaled_left = proof.left_eval * left_scale;
+        let right_scale: E = self.quant_info.right_scale().to_field();
+        let left_claim = Claim::new(last_claim.point.clone(), proof.left_eval);
+        let scaled_right = proof.right_eval * right_scale;
+        let right_claim = Claim::new(last_claim.point.clone(), proof.right_eval);
+        ensure!(
+            scaled_left + scaled_right == last_claim.eval,
+            "Add layer verification failed"
+        );
```

Comment:
Don't we need to take into account also the `global_multiplier` in case no requantization is needed?

Reply:
i think it's taken care of by latest PR fixing accuracy.